File: C++/DesignPattern/CreationalDesignPattern/Decorator_design.cpp
#include <iostream>
#include <memory>
#include <string>
using namespace std;

// Product interface
struct INotification {
    virtual void send(const string& message) = 0;
    virtual ~INotification() = default;
};

struct SMSNotification : INotification {
    void send(const string& message) override {
        cout << "Sending SMS notification " << message << "\n";
    }
};

// Base decorator
class NotificationDecorator : public INotification {
protected:
    shared_ptr<INotification> baseNotification;
public:
    explicit NotificationDecorator(shared_ptr<INotification> notification)
        : baseNotification(move(notification)) {}
};

struct FormatNotification : NotificationDecorator {
    explicit FormatNotification(shared_ptr<INotification> notification)
        : NotificationDecorator(move(notification)) {}

    void send(const string& message) override {
        cout << "Notification is formatted\n";
        baseNotification->send(message);
    }
};

struct AnimationNotification : NotificationDecorator {
    explicit AnimationNotification(shared_ptr<INotification> notification)
        : NotificationDecorator(move(notification)) {}

    void send(const string& message) override {
        cout << "AnimationNotification is added\n";
        baseNotification->send(message);
    }
};

int main() {
    shared_ptr<INotification> notification = make_shared<SMSNotification>();
    notification->send("pradeeep");
    return 0;
}

File: Python/DesignPattern/CreationalDesignPattern/Decorator_design.py
from abc import ABC, abstractmethod


# Product interface
class INotification(ABC):
    @abstractmethod
    def send(self, message: str): pass


class SMSNotification(INotification):
    def send(self, message: str):
        print(f"Sending SMS notification {message}")


# Base decorator
class NotificationDecorator(INotification, ABC):
    def __init__(self, notification: INotification):
        self._base_notification = notification


class FormatNotification(NotificationDecorator):
    def send(self, message: str):
        print("Notification is formatted")
        self._base_notification.send(message)


class AnimationNotification(NotificationDecorator):
    def send(self, message: str):
        print("AnimationNotification is added")
        self._base_notification.send(message)


if __name__ == "__main__":
    notification: INotification = SMSNotification()
    notification.send("pradeeep")
